### PR TITLE
localization: add special type [Experimental]

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
All user-facing strings should be ran through the `localize` function so they can be extracted later for translation. We currently have no way to enforce this programmatically; we must review code and verify that contributors are localizing their strings.

## Solution
Add a special `LocalizedString` type returned by any `localize` call. We can now write functions that accept _only_ a localized string. This PR has move parts of the `messages.ts` file to use this type.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/vscode/localize.d.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as nls from 'vscode-nls'
+import { NominalType } from '../../shared/utilities/tsUtils'
+
+type PatchOverloadedReturnType<T, R> = T extends {
```
Comment (JadenSimon) [inline src/shared/vscode/localize.d.ts:9]: This type only patches the first two overloaded signatures it finds. It's extremely difficult (if it's even possible) to write a type alias to parse an arbitrary number of call signatures (trust me, I tried...)

File: src/shared/utilities/tsUtils.ts [hunk 2]
```diff
@@ -31,3 +31,13 @@ type Expand<T> = T extends infer O ? { [K in keyof O]+?: O[K] } : never
  *
  */
 export type ExpandWithObject<T> = Expand<T> extends Record<string, unknown> ? Expand<T> : never
+
+// Special class to embed a hidden 'key' into a type
+abstract class NominalKey<K> {
```
Comment (JadenSimon) [inline src/shared/utilities/tsUtils.ts:36]: Why this works:
Classes are a special case in TypeScript in that the type-checker will check for _all_ fields, even those marked private/protected. We can then create intersections with arbitrary _T_ types, but on the outside these intersections still look like _T_. But now if you tried to assign _T_ to `T & NominalKey<K>` it will not work because _T_ is missing `__type__: K`. Since the class is abstract, we don't need to assign anything to the field, making the type itself ephemeral. 

Basically:
`T extends (T & NominalKey<K>)` <-- FALSE
`(T & NominalKey<K>) extends T` <-- TRUE
The intersection type is always a valid sub-type of _T_, but the inverse is not true.

**Review by bryceitoc9 — COMMENTED** [review]
Sanity check: is this overengineered for too little gain? I'm seeing some holes here:
* You stated that we can't actually match/enforce the signature. We have some [localized strings that use a LOT of extra params](https://github.com/aws/aws-toolkit-vscode/blob/master/package.nls.json#L377), so there's precedent for more than just a couple signatures.
* https://github.com/aws/aws-toolkit-vscode/pull/2000/files#diff-787292f6a6a82318786a4f99b40c5000995a55e2c13a6ed8447734fa592c9a85R14 shows how easy it is to circumvent this if necessary

What if you just made a class that held localized strings that we replace all `localize` calls with? That provides a few benefits:
* You can type the arg based on the class type
* `vscode-nls` imports will only be needed in one file
* It's more conventional: if I'm joining the team and I saw the Typescript wizardry, I don't think I would understand it.

Both of these approaches also include a significant "good intention" that everyone will use this type going forward. How would we enforce that when we can't already enforce localizations well enough right now?

**Comment by justinmk3** [comment]
> But now if you tried to assign _T_ to `T & NominalKey<K>` it will not work because _T_ is missing `__type__: K`.

IIUC, a wrapper class as @bryceitoc9 mentioned achieves a similar mechanism without adding our own type system overlay.

I also hesitate to introduce something that makes nominal types more attractive. Usually, nominal types are too specific and make code-sharing difficult. Most data is not that special.

**Comment by justinmk3** [comment]
Would a custom lint rule work?
